Sparse index: `git reset`

vdye opened the pull request:
## Changes
In addition to the usual changes a sparse index integration needs, there were some modifications needed to make sure `reset` could handle sparse directory entries and keep the index (especially the cache tree) in a valid state:

- for `reset --mixed`, add `change` and `add_remove` functions in diff settings, as those functions are applied to sparse directories in `oneway_diff` (the unpacking function used by mixed reset)
  - related to this, the `recursive` diff setting is enabled to merge in nested outside-of-cone files
- for `reset --merge` and `reset --hard`, the cache tree is manually reconstructed (from a given tree, _not_ from the index itself, as it is in `cache_tree_update`) in `prime_cache_tree`. In order for the cache tree to be accurate, directory contents are _only_ excluded if that directory exists in the index as a sparse directory
  - this check feels inefficient (each time a directory outside the sparse checkout cone is reached, the entire index is searched for a matching entry). If there's a function that does this search more efficiently, please let me know!
- tests added for all non-`soft` reset modes (since `--soft` doesn't touch the index), include tests ensuring the index isn't expanded

Note: this doesn't address the bug found in the existing sparse reset - if it's alright with you (reviewers), I'll create a separate issue for that.

vdye committed:
sparse-index: integrate with reset

This change enables running the `reset` builtin command without expanding the
full index in a cone-mode sparse checkout.

`reset --soft` does not modify the index, so no compatibility changes are
needed for it to function without expanding the index. For all other reset
modes (`--mixed`, `--hard`, `--keep`, `--merge`), the full index is explicitly
expanded with `ensure_full_index` to maintain current behavior.

Additionally, the `read_cache()` check verifying an uncorrupted index is moved
after argument parsing and preparing the repo settings. The index is not used
by the preceding argument handling, but `read_cache()` does need to be run
after enabling sparse index for the command and before using the index for the
reset.

Signed-off-by: Victoria Dye <vdye@github.com>

vdye committed:
sparse-index: make mixed `reset` sparse-aware

Due to the introduction of sparse directory entries in the `diff_cache`
execution (used internally by `reset --mixed`), explicit `change` and
`add_remove` functions are specified for merging of the interior contents of
those sparse directories.

Additionally, to handle cases in which `reset --mixed` must merge files nested
multiple levels deep in directories outside the sparse checkout cone, the
`recursive` diff option is added.

Signed-off-by: Victoria Dye <vdye@github.com>

derrickstolee commented on cache-tree.c:
```diff
@@ -785,6 +785,23 @@ static void prime_cache_tree_rec(struct repository *r,
 
 	oidcpy(&it->oid, &tree->object.oid);
+
+	/*
+	 * If this entry is a sparse directory with a sparse index, set the entry
+	 * count to 1 and exit without populating subtrees.
+	 * TODO: find a way to make this cleaner/more efficient (i.e., don't loop
+	 * over the index multiple times).
+	 */
+	if (r->index->sparse_index && !path_in_cone_modesparse_checkout(tree_path->buf, r->index)) {
+		for (int i=0; i < r->index->cache_nr; i++) {
+			if (!subtree_name_cmp(r->index->cache[i]->name, r->index->cache[i]->ce_namelen,
+					tree_path->buf, tree_path->len)) {
+				it->entry_count = 1;
+				return;
+			}
+		}
+	}
```
It took me a while to realize why it was so important that we look for the existence of the tree in the index, so I just deleted your loop and set the `entry_count` to 1. The `diff with directory/file conflicts` test failed, catching the problem. The case in that test is that an entry within `folder1/0/0/0` is edited or conflicted, requiring that the cache tree extend further down into `folder1/` even though it is at the sparse-checkout boundary.

I think this will work:

```
	/*
	 * If this entry is outside the sparse-checkout cone, then it might be
	 * a sparse directory entry. Check the index to be sure if it is. It is
	 * possible that a staged or conflicted entry prevents this directory
	 * from collapsing to a sparse directory entry, so look for an entry with
	 * the exact same name as the tree.
	 */
	if (r->index->sparse_index &&
	    !path_in_cone_modesparse_checkout(tree_path->buf, r->index) &&
	    index_name_pos(r->index, tree_path->buf, tree_path->len) >= 0) {
		it->entry_count = 1;
		return;
	}
```

vdye replied:
Perfect, that's exactly what I was looking for!!

ldennington reviewed (commented):
Looks really good to me! I will wait on others who are more familiar with `cache-tree` and the options you added to give their approval, though (I'll admit I got a bit lost with those changes).

derrickstolee commented on t/t1092-sparse-checkout-compatibility.sh:
```diff
@@ -480,6 +480,62 @@ test_expect_success 'checkout and reset (mixed) [sparse]' '
+	test_all_match git reset --hard update-deep &&
+	run_on_sparse test_path_is_missing folder1 &&
+	test_path_exists full-checkout/folder1
+'
```
I appreciate this additional test, documenting how `git reset` behaves in a sparse-checkout scenario. This test could exist on its own at the start of the series (to really show that the sparse index is not involved here) and the NEEDSWORK comment could mention something like "It is possible that we would want to change this behavior to not write files outside of the sparse-checkout cone."

vdye replied:
Done!

derrickstolee commented on t/t1092-sparse-checkout-compatibility.sh:
```diff
@@ -639,15 +695,6 @@ test_expect_success 'submodule handling' '
-	test_region index convert_to_sparse trace2.txt &&
-	test_region index ensure_full_index trace2.txt
-'
-
```
Instead of deleting this test, perhaps we try to find a command that still does the expansion? Perhaps `git read-tree -mu HEAD` would suffice? This is really testing that `command_requires_full_index` works as expected.

vdye replied:
Updated! I tried using some of the other commands that don't yet have sparse index implemented (e.g., 'git diff') but didn't find one and assumed this test was specific to `reset --hard`. Thanks for clarifying (and for the example command)!

derrickstolee commented on t/t1092-sparse-checkout-compatibility.sh:
```diff
@@ -480,6 +480,62 @@ test_expect_success 'checkout and reset (mixed) [sparse]' '
+	test_path_exists full-checkout/folder1
+'
+
+test_expect_success 'checkout and reset (merge)' '
+	init_repos &&
+
+	write_script edit-contents <<-\EOF &&
+	echo text >>$1
+	EOF
+
+	test_all_match git checkout -b reset-test update-deep &&
+	run_on_all ../edit-contents a &&
+	test_all_match git reset --merge deepest &&
+
+	test_all_match git reset --hard update-deep &&
+	run_on_all ../edit-contents deep/a &&
+	test_all_match test_must_fail git reset --merge deepest
+'
+
+test_expect_success 'checkout and reset (keep)' '
+	init_repos &&
+
+	write_script edit-contents <<-\EOF &&
+	echo text >>$1
+	EOF
+
+	test_all_match git checkout -b reset-test update-deep &&
+	run_on_all ../edit-contents a &&
+	test_all_match git reset --keep deepest &&
+
+	test_all_match git reset --hard update-deep &&
+	run_on_all ../edit-contents deep/a &&
+	test_all_match test_must_fail git reset --keep deepest
+'
```
Since these are just further tests of behavior, they could be added in their own commit, mentioning that we want to expand the test coverage of other `git reset` modes.

vdye replied:
These tests have been moved to their own commit (plus three more dealing with pathspec resets). 

I do want to note that, in the mixed reset commit, one of the tests was changed to `test_expect_failure` (specifically, "reset with file pathspec outside sparse definition"). I added the reasoning there for why the test is now failing (similar to the `git blame` test dealing with pathspecs outside the definition). Let me know if you think it's worth trying to correct (e.g., expanding the full index when a pathspec like that is provided).

derrickstolee commented on t/t1092-sparse-checkout-compatibility.sh:
```diff
@@ -680,9 +727,9 @@ test_expect_success 'sparse-index is not expanded' '
 	ensure_not_expanded checkout - &&
 	ensure_not_expanded switch rename-out-to-out &&
 	ensure_not_expanded switch - &&
-	git -C sparse-index reset --hard &&
+	ensure_not_expanded reset --hard &&
 	ensure_not_expanded checkout rename-out-to-out -- deep/deeper1 &&
-	git -C sparse-index reset --hard &&
+	ensure_not_expanded reset --hard &&
```
These changes could be included in the commit that allows `git reset --hard` to work with the sparse index, to demonstrate that the code change is the reason this test change makes sense.

vdye replied:
Updated! Also, noted that tests should generally be embedded in the commits corresponding to their changes (unless the standalone change is "add more tests").

derrickstolee commented on t/t1092-sparse-checkout-compatibility.sh:
```diff
@@ -692,6 +739,13 @@ test_expect_success 'sparse-index is not expanded' '
 	echo >>sparse-index/untracked.txt &&
 	ensure_not_expanded add . &&
 
+	ensure_not_expanded reset --hard &&
+	echo >>sparse-index/README.md &&
+	ensure_not_expanded reset --mixed &&
+	ensure_not_expanded reset --keep update-deep &&
+	ensure_not_expanded reset --merge rename-out-to-out &&
+	ensure_not_expanded reset --hard &&
+
```
These could also be sprinkled into the necessary commits that make their lines possible.

I do notice that your examples for `--mixed` and `--hard` imply that your `<treeish>` is `HEAD`, which might not be the most interesting case. It might be good to do a cycle of the branches that do interesting things.


```
for ref in update-deep update-folder1 update-folder2 update-deep
do
ensure_not_expanded reset --mixed $ref
ensure_not_expanded reset --hard $ref
done &&
```

(I didn't include tabs because of the web editor.)

vdye replied:
Added this test loop for `--hard` and `--mixed` (both added in their relevant commits).

derrickstolee commented on cache-tree.c:
```diff
@@ -777,7 +777,8 @@ int write_index_as_tree(struct object_id *oid, struct index_state *index_state,
 
 static void prime_cache_tree_rec(struct repository *r,
 				 struct cache_tree *it,
-				 struct tree *tree)
+				 struct tree *tree,
+				 struct strbuf *tree_path)
```
These changes to `prime_cache_tree_rec()` seem premature for this commit. We are not using `tree_path` in any interesting capacity yet.

vdye replied:
Thanks for catching this! The `prime_cache_tree_rec` changes are now in the mixed reset commit.

derrickstolee commented on cache-tree.c:
```diff
@@ -791,13 +792,18 @@ static void prime_cache_tree_rec(struct repository *r,
 			sub = cache_tree_sub(it, entry.path);
 			sub->cache_tree = cache_tree();
-			prime_cache_tree_rec(r, sub->cache_tree, subtree);
+			strbuf_addf(&subtree_path, "%s%s/", tree_path->buf, entry.path);
```
`strbuf_addf()` should generally be avoided for performance reasons. This is a really fast path, so I was only seeing ~8ms in `prime_cache_tree()` even for a full monorepo without sparse index, but let's use our best tricks when possible.

See derrickstolee/git@bf2a9d3 for the fixup.

vdye replied:
Pulled and squashed the fixup (and removed the extra `struct strbuf subtree_path = STRBUF_INIT;`).

derrickstolee reviewed (commented):
I took a look through the code and left some comments that are mostly about crafting a good patch series. The biggest one is to make your commits as focused as possible: do only one thing at a time. But that _also_ means, demonstrate how that one thing changes functionality through tests, if possible. The test dump at the end should be replaced with a sprinkling of tests throughout.

The tests could probably use some cases that specify a pathspec, both in and outside of the sparse-checkout cone.

I installed your version of Git and tried a bunch of `git reset` shenanigans to see if I could trigger an index expansion or other slow behavior. I did see a really slow (5s) loop inside `unpack_trees()` during a `git reset --hard -- <path>`, but that's unrelated to the sparse index. It happens to be the loop conditioned by this `if` in `unpack_trees()`:

```c
	/* Any left-over entries in the index? */
	if (o->merge) {
```

Turns out that this loops through all of the index entries in a very slow way. This might be something to investigate separately (and much later than the current work).

vdye committed:
sparse-index: make non-mixed `reset` sparse-aware

This change allows non-mixed resets to no longer require that the full index
is expanded. This is accomplished by updating the `prime_cache_tree_rec`
function to have it reconstruct the cache tree "aware" of whether a directory
is sparse in the index or not. This check must be done by verifying the
contents of the cache itself (rather than checking whether a directory is
inside of the sparse checkout cone) as entries may have been added to the
sparse index outside of the checkout cone.

Signed-off-by: Victoria Dye <vdye@github.com>

vdye committed:
reset: add test demonstrating mixed reset outside sparse checkout

Mixed reset in a sparse checkout (cone mode or otherwise) will, if a file
outside of the sparse definition has differences between the working tree
and target, write the file to the local working copy of the repository.
Since this behavior conflicts with the typical expectations of a strict
checkout definition/cone, a test is added demonstrating (and explaining) it.

Signed-off-by: Victoria Dye <vdye@github.com>

vdye committed:
sparse-index: use `read-tree` for expand/collapse test

In anticipation of `reset` being fully integrated with sparse index, the
test for index expand and collapse for non-sparse index integrated commands
is updated to use `read-tree`.

Signed-off-by: Victoria Dye <vdye@github.com>

vdye committed:
reset: expand test coverage in sparse checkout

Add tests for `--merge` and `--keep` modes, as well as (mixed) reset with
pathspecs both inside and outside of the sparse checkout definition.

Signed-off-by: Victoria Dye <vdye@github.com>

derrickstolee commented on builtin/reset.c:
```diff
@@ -281,9 +282,6 @@ static void parse_args(struct pathspec *pathspec,
 	}
 	*rev_ret = rev;
 
-	if (read_cache() < 0)
-		die(_("index file corrupt"));
-
```
In this commit, it might be worth mentioning the movement of this check. It is fine to move until after finishing the parsing of options, but the `command_requires_full_index` check needs to be before `read_cache()`.

vdye replied:
Details added to the commit message.

derrickstolee commented on t/t1092-sparse-checkout-compatibility.sh:
```diff
@@ -562,7 +562,12 @@ test_expect_success 'reset with sparse directory pathspec outside definition' '
 '
 
-test_expect_success 'reset with file pathspec outside sparse definition' '
+# TODO: reset with a file pathspec outside of the sparse checkout definition
+# does not currently work in cone-mode checkouts. This is due to the tree's
+# directory entry (that contains the file) not being expanded before the
+# pathspec check is performed (in the case of this test, "folder1/" is compared
+# to the pathspec "folder1/a", and does not match).
```
Thanks for pointing out how this is failing. We should debug into this tomorrow to see what logic is happening. Something like `index_name_pos(istate, "folder1/a")` would expand a sparse index, so something different must be happening. Perhaps it's related to how `reset` uses the cache-tree structure.

vdye replied:
I did find the the source of the failure, but wasn't sure whether I should have pointed out the specifics in the `TODO` comment: https://github.com/microsoft/git/blob/vfs-2.33.0/diff-lib.c#L505-L507

In the linked `ce_path_match` check, `idx->name` is `folder1/` (since it's initially a sparse entry), but `revs->prune_data` (set to the input pathspec) is `folder1/a`. The `ce_path_match` check fails, so `do_oneway_diff` is never called and that part of the tree is effectively ignored. This also explains why a sparse directory pathspec (like `folder1`) _does_ work - the match succeeds, and the relevant entries are expanded in `do_oneway_diff`.

The part I'm not sure about is the right way to fix it (if it should be fixed at all). Should the full index be expanded right at the beginning in `builtin/reset.c` if the pathspec is outside the sparse checkout cone (although how would you handle a wildcard pathspec like `*.c`)? Or should pathspec usage be blocked by default, but expanding the full index would be allowed using a `--sparse` input flag? There are definitely other ways of doing it, but I imagine we probably want something that doesn't preemptively expand index entries if we can avoid it.

derrickstolee replied:
The fact that we are getting into `oneway_diff()` with a `folder1/` entry an a `folder1/a` entry makes me think that we need more updates to `unpack_trees()` to catch this case during that update. When you debug into that line, is there anything in the stack trace that looks like a good place to identify that we need to expand the index here?

As for preventing the user request, there is some precedent in `builtin/add.c` using `matches_skip_worktree()` and `path_in_sparse_checkout()` that might be illuminating. Take a look and see if you are inspired for an approach in `git reset`. There is absolutely momentum building to prevent these kinds of interactions outside of the sparse-checkout cone.

vdye replied:
Sorry, I should clarify - `folder1/a` isn't an entry (as in a tree or cache entry), it's a `pathspec` pulled directly from command-line input that's compared to `idx` (a sparse directory cache entry). Both `idx` and `tree` (the two cache entries compared in `oneway_diff`) are sparse directory entries pointing at `folder1/`. The only indicator of needing to expand the index before that (as far as I can tell) is the pathspec itself, which could be used as early as `cmd_reset`. 

I'll check out `matches_skip_worktree()` and `path_in_sparse_checkout()` - if they can interpret wildcards in the pathspec (or can be updated to do that), it should be straightforward to expand the index before even entering `unpack_trees()`

vdye replied:
I've just added two fixups (for readability in this PR - I'll squash them before merging) that should fix this. The first adds more tests for pathspec resets, and the second adds checks to expand the index if the pathspec isn't obviously for things in the index (and corresponding tests for when the index should _not_ be expanded).

derrickstolee reviewed (commented):
Great job reorganizing the commits!

derrickstolee commented on cache-tree.c:
```diff
@@ -777,13 +777,32 @@ int write_index_as_tree(struct object_id *oid, struct index_state *index_state,
+		it->entry_count = 1;
+		return;
+	}
+
```
nit: double newline.

derrickstolee commented on t/t1092-sparse-checkout-compatibility.sh:
```diff
@@ -814,8 +809,18 @@ test_expect_success 'sparse-index is not expanded' '
+	ensure_not_expanded reset base -- nonexistent-file &&
+	ensure_not_expanded reset deepest -- deep &&
+
+	# Although folder1 is outside the sparse definition, it exists as a
+	# directory entry in the index, so it will be reset without needing to
+	# expand the full index.
+	ensure_not_expanded reset --hard update-folder1 &&
```
nice! I'm excited to see that this is possible.

derrickstolee reviewed (approved):
I have a tiny nit for you to fixup while you rebase. Otherwise this looks great!
